File: dyros_robot_menagerie/dyros_robot_menagerie/husky/controller.py
import math
from typing import Dict
import numpy as np
from scipy.spatial.transform import Rotation as R
from rclpy.node import Node
from std_msgs.msg import Int32
from geometry_msgs.msg import PoseStamped, Twist
from sensor_msgs.msg import JointState
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy
from mujoco_ros_sim import ControllerInterface
from dyros_robot_menagerie.husky.robot_data import (HuskyRobotData,
                                                    TASK_DOF,
                                                    WHEEL_DOF,
                                                    )
from drc.mobile.robot_controller import RobotController

"""
MuJoCo Model Information: husky
 id | name                 | type   | nq | nv | idx_q | idx_v
----+----------------------+--------+----+----+-------+------
  0 | -                    | Free   |  7 |  6 |     0 |    0
  1 | front_left_wheel     | Hinge  |  1 |  1 |     7 |    6
  2 | front_right_wheel    | Hinge  |  1 |  1 |     8 |    7
  3 | rear_left_wheel      | Hinge  |  1 |  1 |     9 |    8
  4 | rear_right_wheel     | Hinge  |  1 |  1 |    10 |    9

 id | name                 | trn     | target_joint
----+----------------------+---------+-------------
  0 | left_wheel           | Joint   | front_left_wheel
  1 | right_wheel          | Joint   | front_right_wheel

 id | name                        | type             | dim | adr | target (obj)
----+-----------------------------+------------------+-----+-----+----------------
  0 | position_sensor             | FramePos         |   3 |   0 | Site:husky_site
  1 | orientation_sensor          | FrameQuat        |   4 |   3 | Site:husky_site
  2 | linear_velocity_sensor      | FrameLinVel      |   3 |   7 | Site:husky_site
  3 | angular_velocity_sensor     | FrameAngVel      |   3 |  10 | Site:husky_site

 id | name                        | mode     | resolution
----+-----------------------------+----------+------------
  0 | front_view                  | -        | 640x480
  1 | left_view                   | -        | 640x480
  2 | right_view                  | -        | 640x480
  3 | upper_view                  | -        | 640x480
"""
class HuskyControllerPy(ControllerInterface):
    def __init__(self, node: Node):
        super().__init__(node)

        self.dt = 0.01
        
        # robot data & controller
        self.robot_data       = HuskyRobotData()
        self.robot_controller = RobotController(self.dt, self.robot_data)
        
        # ROS 2 I/O                                                          #
        qos = QoSProfile(depth=1)
        qos.reliability = ReliabilityPolicy.BEST_EFFORT
        qos.history = HistoryPolicy.KEEP_LAST

        ns = "husky_controller"
        self.key_sub   = node.create_subscription(Int32,       f"{ns}/mode_input",  self._key_cb,         10)
        self.pose_sub  = node.create_subscription(PoseStamped, f"{ns}/target_pose", self._pose_cb,        10)
        self.vel_sub   = node.create_subscription(Twist,       f"{ns}/cmd_vel",     self._vel_cb,         10)
        self.joint_sub = node.create_subscription(JointState,  "/joint_states_raw", self._joint_state_cb, 10)

        self.pose_pub  = node.create_publisher(PoseStamped, f"{ns}/base_pose",  10)
        self.vel_pub   = node.create_publisher(Twist,       f"{ns}/base_vel",   10)
        self.joint_pub = node.create_publisher(JointState,  "/joint_states",    10)
        self.front_image_pub = node.create_publisher(Image, f'{ns}/front_view', qos)
        self.left_image_pub  = node.create_publisher(Image, f'{ns}/left_view',  qos)
        
        # Image converter
        self.bridge_ = CvBridge()

        # states
        self.mode              = "Stop"
        self.is_mode_changed   = True
        self.goal_pose_changed = False

        self.current_time        = 0.0
        self.control_start_time  = 0.0
        
        self.base_pose         = np.zeros(TASK_DOF)
        self.base_pose_desired = np.zeros(TASK_DOF)
        self.base_pose_init    = np.zeros(TASK_DOF)
        self.base_vel          = np.zeros(TASK_DOF)
        self.base_vel_init     = np.zeros(TASK_DOF)
        self.base_vel_desired  = np.zeros(TASK_DOF)

        self.base_vel_world    = np.zeros(TASK_DOF)
        
        self.wheel_pos         = np.zeros(WHEEL_DOF)
        self.wheel_vel         = np.zeros(WHEEL_DOF)
        
        self.wheel_vel_desired = np.zeros(WHEEL_DOF)
        
        # Image
        self.front_image = None
        self.left_image = None
        
        lines = [
            "=================================================================",
            "=================================================================",
            "URDF Joint Information: Husky",
            self.robot_data.get_verbose().rstrip("\n"),
            "=================================================================",
            "=================================================================",
        ]
        text = "\n".join(lines)
        self.node.get_logger().info("\033[1;34m\n" + text + "\033[0m")

    def starting(self) -> None:
        self.pose_timer = self.node.create_timer(0.1, self._pub_pose_cb)
        self.vel_timer  = self.node.create_timer(0.1, self._pub_vel_cb)
        self.front_image_timer = self.node.create_timer(0.01, self._pub_front_image_cb)
        self.left_image_timer = self.node.create_timer(0.01, self._pub_left_image_cb)

    def updateState(self,
                    pos_dict: Dict[str, np.ndarray],
                    vel_dict: Dict[str, np.ndarray],
                    tau_ext_dict: Dict[str, np.ndarray],
                    sensor_dict: Dict[str, np.ndarray],
                    current_time: float,
                    ) -> None:
        self.current_time = current_time

        # get mobile wheel joint
        self.wheel_pos[:] = np.array([pos_dict["front_left_wheel"][0],
                                      pos_dict["front_right_wheel"][0]])
        self.wheel_vel[:] = np.array([vel_dict["front_left_wheel"][0],
                                      vel_dict["front_right_wheel"][0]])

        # get base pose (world)
        self.base_pose[:2] = sensor_dict["position_sensor"][:2]
        self.base_pose[2]  = R.from_quat(sensor_dict["orientation_sensor"], scalar_first=True).as_euler('zyx', degrees=False)[0]

        # base velocity  (world, base)
        self.base_vel_world[:2] = sensor_dict["linear_velocity_sensor"][:2]
        self.base_vel_world[2] = sensor_dict["angular_velocity_sensor"][2]
        rot = np.array([[ math.cos(self.base_pose[2]),  math.sin(self.base_pose[2])],
                        [-math.sin(self.base_pose[2]),  math.cos(self.base_pose[2])]])
        self.base_vel[:2] = rot @ self.base_vel_world[:2]
        self.base_vel[2]  = self.base_vel_world[2]

        # update internal robot model
        if not self.robot_data.update_state(self.wheel_pos, self.wheel_vel):
            self.node.get_logger().error("[HuskyRobotData] Failed to update robot state.")
            
    def updateRGBDImage(self, rgbd_dict: Dict[str, np.ndarray]) -> None:
        self.front_image = rgbd_dict.get("front_view", {}).get("rgb")
        self.left_image = rgbd_dict.get("left_view", {}).get("rgb")

    def compute(self) -> None:
        if self.mode_changed:
            self.mode_changed = False
            self.control_start_time = self.current_time
            
            self.base_pose_init      = self.base_pose.copy()
            self.base_pose_desired   = self.base_pose_init.copy()
            self.base_vel_init       = self.base_vel.copy()
            self.base_vel_desired[:] = 0.0

        if self.mode == "STOP":
            self.wheel_vel_desired[:] = 0.0
        elif self.mode == "Base Velocity Tracking":
            self.wheel_vel_desired = self.robot_controller.velocity_command(self.base_vel_desired)
        else:
            self.wheel_vel_desired[:] = 0.0

    def getCtrlInput(self) -> Dict[str, float]:
        return {"left_wheel":  self.wheel_vel_desired[0],
                "right_wheel": self.wheel_vel_desired[1],
                }

    def _set_mode(self, mode: str) -> None:
        self.mode = mode
        self.mode_changed = True
        self.node.get_logger().info(f"Mode changed: {mode}")

    def _key_cb(self, msg: Int32) -> None:
        mapping = {
            1: "STOP",
            2: "Base Velocity Tracking",
        }
        self._set_mode(mapping.get(msg.data, "NONE"))

    def _pose_cb(self, msg: PoseStamped):
        self.base_pose_desired[:2] = [msg.pose.position.x, msg.pose.position.y]
        self.base_pose_desired[2]  = R.from_quat(np.array([msg.pose.orientation.x,
                                                           msg.pose.orientation.y,
                                                           msg.pose.orientation.z,
                                                           msg.pose.orientation.w]), 
                                                 scalar_first=True).as_euler('zyx', degrees=False)[2]

    def _vel_cb(self, msg: Twist):
        self.base_vel_desired[:] = [msg.linear.x, msg.linear.y, msg.angular.z]

    def _joint_state_cb(self, msg: JointState):
        js = JointState()
        js.header   = msg.header
        js.name     = list(msg.name)
        js.position = list(msg.position)
        js.velocity = list(msg.velocity)
        js.effort   = list(msg.effort)

        js.name.extend(["v_x_joint", "v_y_joint", "v_t_joint"])
        js.position.extend(self.base_pose.tolist())
        js.velocity.extend(self.base_vel_world.tolist())
        js.effort.extend([0.0, 0.0, 0.0])

        self.joint_pub.publish(js)

    def _pub_pose_cb(self):
        msg = PoseStamped()
        msg.header.frame_id = "world"
        msg.header.stamp    = self.node.get_clock().now().to_msg()
        msg.pose.position.x, msg.pose.position.y = self.base_pose[:2]
        qz = math.sin(self.base_pose[2] / 2.0)
        qw = math.cos(self.base_pose[2] / 2.0)
        msg.pose.orientation.z = qz
        msg.pose.orientation.w = qw
        self.pose_pub.publish(msg)

    def _pub_vel_cb(self):
        msg = Twist()
        msg.linear.x, msg.linear.y, msg.angular.z = self.base_vel
        self.vel_pub.publish(msg)

    def _pub_front_image_cb(self):
        if self.front_image is None:
            self.node.get_logger().warn("front_image not ready yet")
            return
        img = np.ascontiguousarray(self.front_image)

        img_bgr = img[:, :, ::-1]
        ros_image = self.bridge_.cv2_to_imgmsg(img_bgr, encoding='bgr8')

        ros_image.header.stamp = self.node.get_clock().now().to_msg()
        ros_image.header.frame_id = 'front_view'
        self.front_image_pub.publish(ros_image)
        
    def _pub_left_image_cb(self):
        if self.left_image is None:
            self.node.get_logger().warn("left_image not ready yet")
            return
        img = np.ascontiguousarray(self.left_image)

        img_bgr = img[:, :, ::-1]
        ros_image = self.bridge_.cv2_to_imgmsg(img_bgr, encoding='bgr8')

        ros_image.header.stamp = self.node.get_clock().now().to_msg()
        ros_image.header.frame_id = 'left_view'
        self.left_image_pub.publish(ros_image)

File: dyros_robot_menagerie/src/husky/controller.cpp
#include "dyros_robot_menagerie/husky/controller.h"

namespace Husky
{
    void HuskyController::configure(const rclcpp::Node::SharedPtr& node)
    {
        MujocoRosSim::ControllerInterface::configure(node);
        dt_ = 0.01;
        
        robot_data_ = std::make_shared<HuskyRobotData>();
        robot_controller_ = std::make_unique<drc::Mobile::RobotController>(dt_, robot_data_);

        key_sub_ = node_->create_subscription<std_msgs::msg::Int32>("husky_controller/mode_input", 10,std::bind(&HuskyController::keyCallback, this, std::placeholders::_1));
        target_pose_sub_ = node_->create_subscription<geometry_msgs::msg::PoseStamped>("husky_controller/target_pose", 10,std::bind(&HuskyController::subtargetPoseCallback, this, std::placeholders::_1));
        target_vel_sub_ = node_->create_subscription<geometry_msgs::msg::Twist>("husky_controller/cmd_vel", 10,std::bind(&HuskyController::subtargetVelCallback, this, std::placeholders::_1));
        joint_sub_ = node_->create_subscription<sensor_msgs::msg::JointState>("/joint_states_raw", 10, std::bind(&HuskyController::subJointStatesCallback, this, std::placeholders::_1));
        
        base_pose_pub_ = node_->create_publisher<geometry_msgs::msg::PoseStamped>("husky_controller/base_pose", 10);
        base_vel_pub_ = node_->create_publisher<geometry_msgs::msg::Twist>("husky_controller/base_vel", 10);
        joint_pub_ = node_->create_publisher<sensor_msgs::msg::JointState>("/joint_states", 10);
        
        base_pose_.setZero();
        base_pose_desired_.setZero();
        base_pose_init_.setZero();

        base_vel_.setZero();
        base_vel_tmp.setZero();
        base_vel_desired_.setZero();
        base_vel_init_.setZero();

        wheel_vel_desired_.setZero();

        std::ostringstream oss;
        oss << "\n=================================================================\n"
            << "=================================================================\n"
            << "URDF Joint Information: Husky\n"
            << robot_data_->getVerbose()
            << "=================================================================\n"
            << "=================================================================";
        const std::string print_info = oss.str();
        RCLCPP_INFO(node->get_logger(), "%s%s%s", cblue, print_info.c_str(), creset);
    }

    void HuskyController::starting()
    {
        base_pose_pub_timer_ = node_->create_wall_timer(std::chrono::milliseconds(100), std::bind(&HuskyController::pubBasePoseCallback, this));
        base_vel_pub_timer_ = node_->create_wall_timer(std::chrono::milliseconds(100), std::bind(&HuskyController::pubBaseVelCallback, this));
    }

    void HuskyController::updateState(const MujocoRosSim::VecMap& pos_dict, 
                                      const MujocoRosSim::VecMap& vel_dict,
                                      const MujocoRosSim::VecMap& tau_ext_dict, 
                                      const MujocoRosSim::VecMap& sensors_dict, 
                                      double current_time)
    {
        current_time_ = current_time;

        // get mobile wheel joint
        wheel_pos_(0) = pos_dict.at("front_left_wheel")(0);
        wheel_pos_(1) = pos_dict.at("front_right_wheel")(0);

        wheel_vel_(0) = vel_dict.at("front_left_wheel")(0);
        wheel_vel_(1) = vel_dict.at("front_right_wheel")(0);

        // get base pose
        base_pose_.head(2) = sensors_dict.at("position_sensor").head(2);
        Quaterniond quat(sensors_dict.at("orientation_sensor")(0),
                         sensors_dict.at("orientation_sensor")(1),
                         sensors_dict.at("orientation_sensor")(2),
                         sensors_dict.at("orientation_sensor")(3));
        Vector3d euler_rpy = DyrosMath::rot2Euler(quat.toRotationMatrix());
        base_pose_(2) = euler_rpy(2);

        // get base vel
        Matrix2d rot_base2world;
        rot_base2world << cos(base_pose_(2)), sin(base_pose_(2)),
                         -sin(base_pose_(2)), cos(base_pose_(2));

        base_vel_.head(2) = rot_base2world * sensors_dict.at("linear_velocity_sensor").head(2);
        base_vel_(2) = sensors_dict.at("angular_velocity_sensor")(2);

        base_vel_tmp.head(2) = sensors_dict.at("linear_velocity_sensor").head(2);
        base_vel_tmp(2) = sensors_dict.at("angular_velocity_sensor")(2);
        
        robot_data_->updateState(wheel_pos_, wheel_vel_);
    }

    void HuskyController::compute()
    {
        if(is_mode_changed_)
        {
            is_mode_changed_ = false;
            control_start_time_ = current_time_;

            base_pose_init_ = base_pose_;
            base_vel_init_ = base_vel_;
            base_pose_desired_ = base_pose_init_;
            base_vel_init_.setZero();
        }
        
        if(mode_ == "Stop")
        {
            wheel_vel_desired_.setZero();
        }
        else if(mode_ == "Base Velocity Tracking")
        {
            wheel_vel_desired_ = robot_controller_->VelocityCommand(base_vel_desired_);
        }
        else
        {
            wheel_vel_desired_.setZero();
        }
    }


    MujocoRosSim::CtrlInputMap HuskyController::getCtrlInput() const
    {
        MujocoRosSim::CtrlInputMap ctrl_dict;
        ctrl_dict["left_wheel"] = wheel_vel_desired_(0);
        ctrl_dict["right_wheel"] = wheel_vel_desired_(1);
        return ctrl_dict;
    }

     void HuskyController::setMode(const std::string& mode)
    {
        is_mode_changed_ = true;
        mode_ = mode;
        RCLCPP_INFO(node_->get_logger(), "\033[34m Mode changed: %s\033[0m", mode.c_str());
    }

    void HuskyController::keyCallback(const std_msgs::msg::Int32::SharedPtr msg)
    {
        RCLCPP_INFO(node_->get_logger(), "Key input received: %d", msg->data);
        if(msg->data == 1)      setMode("Stop");
        else if(msg->data == 2) setMode("Base Velocity Tracking");
        else                    setMode("None");
    }

    void HuskyController::subtargetPoseCallback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
    {
        RCLCPP_INFO(node_->get_logger(),
                    "Target pose received: position=(%.3f, %.3f, %.3f), "
                    "orientation=(%.3f, %.3f, %.3f, %.3f)",
                    msg->pose.position.x, msg->pose.position.y, msg->pose.position.z,
                    msg->pose.orientation.x, msg->pose.orientation.y,
                    msg->pose.orientation.z, msg->pose.orientation.w);
        is_goal_pose_changed_ = true;

        // Convert to 4x4 homogeneous transform
        Eigen::Quaterniond quat(msg->pose.orientation.w,
                                msg->pose.orientation.x,
                                msg->pose.orientation.y,
                                msg->pose.orientation.z);

        Vector3d euler_rpy = DyrosMath::rot2Euler(quat.toRotationMatrix());
        base_pose_desired_(2) = euler_rpy(2);
        base_pose_desired_.head(2) << msg->pose.position.x,
                                       msg->pose.position.y;
    }

    void HuskyController::subtargetVelCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
    {
        RCLCPP_INFO(node_->get_logger(),
                    "Target velocity received: linear=(%.3f, %.3f, %.3f), "
                    "angular=(%.3f, %.3f, %.3f)",
                    msg->linear.x, msg->linear.y, msg->linear.z,
                    msg->angular.x, msg->angular.y, msg->angular.z);

        base_vel_desired_.head(2) << msg->linear.x, msg->linear.y;
        base_vel_desired_(2) = msg->angular.z;
    }

    void HuskyController::subJointStatesCallback(const sensor_msgs::msg::JointState::SharedPtr msg)
    {
        auto joint_msg = sensor_msgs::msg::JointState();
        joint_msg.header = msg->header;
        joint_msg.name = msg->name;
        joint_msg.position = msg->position;
        joint_msg.velocity = msg->velocity;
        joint_msg.effort = msg->effort;

        const std::vector<std::string> virtual_joints = {"v_x_joint", "v_y_joint", "v_t_joint"};
        std::vector<double> virtual_pos_vec(base_pose_.data(), base_pose_.data() + base_pose_.size());
        std::vector<double> virtual_vel_vec(base_vel_tmp.data(), base_vel_tmp.data() + base_vel_tmp.size());
        std::vector<double> virtual_eff_vec{0,0,0};

        joint_msg.name.insert(joint_msg.name.end(), virtual_joints.begin(), virtual_joints.end());
        joint_msg.position.insert(joint_msg.position.end(), virtual_pos_vec.begin(), virtual_pos_vec.end());
        joint_msg.velocity.insert(joint_msg.velocity.end(), virtual_vel_vec.begin(), virtual_vel_vec.end());
        joint_msg.effort.insert(joint_msg.effort.end(), virtual_eff_vec.begin(), virtual_eff_vec.end());

        joint_pub_->publish(joint_msg);
    }

    void HuskyController::pubBasePoseCallback()
    {
        auto base_pose_msg = geometry_msgs::msg::PoseStamped();
        base_pose_msg.header.frame_id = "world";
        base_pose_msg.header.stamp = node_->now();

        base_pose_msg.pose.position.x = base_pose_(0);
        base_pose_msg.pose.position.y = base_pose_(1);
        base_pose_msg.pose.position.z = 0;

        Eigen::Quaterniond quat(Eigen::AngleAxisd(base_pose_(2), Eigen::Vector3d::UnitZ()));
        base_pose_msg.pose.orientation.x = quat.x();
        base_pose_msg.pose.orientation.y = quat.y();
        base_pose_msg.pose.orientation.z = quat.z();
        base_pose_msg.pose.orientation.w = quat.w();
        
        base_pose_pub_->publish(base_pose_msg);
    }

    void HuskyController::pubBaseVelCallback()
    {
        auto base_vel_msg = geometry_msgs::msg::Twist();
        base_vel_msg.linear.x = base_vel_(0);
        base_vel_msg.linear.y = base_vel_(1);
        base_vel_msg.angular.z = base_vel_(2);
        
        base_vel_pub_->publish(base_vel_msg);
    }

    /* register with the global registry */
    PLUGINLIB_EXPORT_CLASS(Husky::HuskyController, MujocoRosSim::ControllerInterface)
} // namespace Husky